File: Reqnroll/Chapter13/13.1-TheLocalScenarioContextPattern/Before/WIMP.LocalScenarioContextSample.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.LocalScenarioContextSample.App.Infrastructure;
using WIMP.LocalScenarioContextSample.App.Services;

namespace WIMP.LocalScenarioContextSample.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions
{
    [Given("the customer {word} has placed the order #{int}")]
    public void GivenTheCustomerHasPlacedTheOrder(string customerName, int orderNo)
    {
        OrderService.PlaceOrder(customerName, orderNo, "Margherita");
    }

    [When("the customer {word} cancels the order #{int}")]
    public void WhenTheCustomerCancelsTheOrder(string customerName, int orderNo)
    {
        OrderService.CancelOrder(customerName, orderNo);
    }

    [Then("the customer {word} should receive a notification about the cancellation")]
    public void ThenTheCustomerShouldReceiveANotification(string customerName)
    {
        Assert.IsTrue(NotificationService.WasNotificationSent(customerName));
    }

    #region Reset database for every scenario execution

    /// <summary>
    /// This hook resets the in-memory database before each scenario execution,
    /// ensuring that each test starts with a clean state.
    /// The pattern TODO1 contains a better approach for in-memory database management via dependencies.
    /// The pattern TODO2 contains further options for dealing with shared resources, such as using a real database.
    /// </summary>
    [BeforeScenario]
    public void ResetDatabase()
    {
        DataContext.Instance.Reset();
    }

    #endregion
}

File: CucumberJVM/Chapter13/13.1-LocalScenarioContextPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.infrastructure.DataRepository;
import com.wimp.app.services.NotificationService;
import com.wimp.app.services.OrderService;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class OrderingStepDefinitions {
    @Given("the customer {string} has placed the order #{int}")
    public void theCustomerHasPlacedTheOrder(String customerName, int orderNo) {
        OrderService.placeOrder(customerName, orderNo, "Margherita");
    }

    @When("the customer {string} cancels the order #{int}")
    public void theCustomerCancelsTheOrder(String customerName, int orderNo) {
        OrderService.cancelOrder(customerName, orderNo);
    }

    @Then("the customer {string} should receive a notification about the cancellation")
    public void theCustomerShouldReceiveANotification(String customerName) {
        assertTrue(NotificationService.wasNotificationSent(customerName));
    }

    /**
     * This hook resets the in-memory database before each scenario execution,
     * ensuring that each test starts with a clean state.
     * The samples of chapter 14 demonstrate an improved approach by having a
     * stub implementation of the data repository and the samples of chapter 15
     * and later further improve that by using H2 in-memory database as stub at
     * Spring database layer. Finally, chapter 18 demonstrates how testing can
     * be done with real databases.
     */
    @Before
    public void resetDatabase() {
        DataRepository.INSTANCE.reset();
    }
}
